packages/seedless-onboarding-controller/src/SeedlessOnboardingController.ts

```diff
@@ -0,0 +1,829 @@
+      parsedVaultData = JSON.parse(data);
+    } catch {
+      throw new Error(SeedlessOnboardingControllerError.InvalidVaultData);
+    }
+
+    this.#assertIsValidVaultData(parsedVaultData);
+
+    const rawToprfEncryptionKey = base64ToBytes(
+      parsedVaultData.toprfEncryptionKey,
+    );
+    const parsedToprfAuthKeyPair = JSON.parse(parsedVaultData.toprfAuthKeyPair);
+    const rawToprfAuthKeyPair = {
+      sk: BigInt(parsedToprfAuthKeyPair.sk),
+      pk: base64ToBytes(parsedToprfAuthKeyPair.pk),
+    };
+
+    return {
+      nodeAuthTokens: parsedVaultData.authTokens,
+      toprfEncryptionKey: rawToprfEncryptionKey,
+      toprfAuthKeyPair: rawToprfAuthKeyPair,
+    };
+  }
+
+  /**
+   * Check if the provided value is a valid node auth tokens.
+   *
+   * @param value - The value to check.
+   * @throws If the value is not a valid node auth tokens.
+   */
+  #assertIsValidNodeAuthTokens(
+    value: unknown,
+  ): asserts value is NodeAuthTokens {
+    if (!Array.isArray(value) || value.length === 0) {
+      throw new Error(SeedlessOnboardingControllerError.InsufficientAuthToken);
+    }
+  }
+
+  /**
+   * Check if the provided value is a valid vault data.
+   *
+   * @param value - The value to check.
+   * @throws If the value is not a valid vault data.
+   */
+  #assertIsValidVaultData(value: unknown): asserts value is VaultData {
+    // value is not valid vault data if any of the following conditions are true:
+    if (
+      !value || // value is not defined
+      typeof value !== 'object' || // value is not an object
+      !('authTokens' in value) || // authTokens is not defined
+      typeof value.authTokens !== 'object' || // authTokens is not an object
+      !('toprfEncryptionKey' in value) || // toprfEncryptionKey is not defined
+      typeof value.toprfEncryptionKey !== 'string' || // toprfEncryptionKey is not a string
+      !('toprfAuthKeyPair' in value) || // toprfAuthKeyPair is not defined
+      typeof value.toprfAuthKeyPair !== 'string' // toprfAuthKeyPair is not a string
+    ) {
+      throw new Error(SeedlessOnboardingControllerError.VaultDataError);
+    }
+  }
+}
+
+/**
+ * Assert that the provided password is a valid non-empty string.
+ *
+ * @param password - The password to check.
+ * @throws If the password is not a valid string.
+ */
+function assertIsValidPassword(password: unknown): asserts password is string {
+  if (typeof password !== 'string') {
+    throw new Error(SeedlessOnboardingControllerError.WrongPasswordType);
+  }
+
+  if (!password || !password.length) {
+    throw new Error(SeedlessOnboardingControllerError.InvalidEmptyPassword);
+  }
+}
+
+/**
+ * Lock the given mutex before executing the given function,
+ * and release it after the function is resolved or after an
+ * error is thrown.
+ *
+ * @param mutex - The mutex to lock.
+ * @param callback - The function to execute while the mutex is locked.
+ * @returns The result of the function.
+ */
+async function withLock<Result>(
```

Comment:
We should really extract this to `controller-utils`. This is the third place I think we've used this.

Oh well, something to do later...